### PR TITLE
Daedalean coding standards for preprocessor directives

### PR DESCRIPTION
**Pull request by finomen**
Only following uses of preprocessor directives are allowed:

- #pragma once MUST be used to prevent multiple inclusions of the same file
- #include <path> MUST be used to include headers

Relates: T10098

### CODE REVIEW

File: clang-tools-extra/clang-tidy/daedalean/PreprocessingDirectivesCheck.cpp [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+
+using namespace clang::ast_matchers;
+
+namespace clang {
+namespace tidy {
+namespace daedalean {
+
+namespace {
+
+enum class GroupType {
```
Comment (aburgm) [inline clang-tools-extra/clang-tidy/daedalean/PreprocessingDirectivesCheck.cpp]: unused type

File: clang-tools-extra/clang-tidy/daedalean/PreprocessingDirectivesCheck.cpp [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+public:
+  explicit IncludeOrderPPCallbacks(ClangTidyCheck &Check,
+                                   const SourceManager &SM)
+      : Check(Check), SM(SM) {}
+
+  virtual void PragmaDirective(SourceLocation Loc,
+                               PragmaIntroducerKind Introducer) override {
+    const auto fileId = SM.getFileID(Loc);
+    const auto fileName = SM.getFileEntryForID(fileId)->getName();
+    if (fileName.endswith(".hpp") || fileName.endswith(".h")) {
```
Comment (aburgm) [inline clang-tools-extra/clang-tidy/daedalean/PreprocessingDirectivesCheck.cpp]: Our headers will have `.hh` extension... where are .hpp and .h coming from?

File: clang-tools-extra/clang-tidy/daedalean/PreprocessingDirectivesCheck.cpp [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+  virtual void Elifndef(SourceLocation Loc, const Token &MacroNameTok,
+                        const MacroDefinition &MD) override {
+    Check.diag(Loc, "Preprocessor directives must not be used");
+  }
+
+  virtual void Elifndef(SourceLocation Loc, SourceRange ConditionRange,
+                        SourceLocation IfLoc) override {
+    Check.diag(Loc, "Preprocessor directives must not be used");
+  }
```
Comment (aburgm) [inline clang-tools-extra/clang-tidy/daedalean/PreprocessingDirectivesCheck.cpp:118]: Should we also catch `#error` and `#line`?

File: clang-tools-extra/clang-tidy/daedalean/PreprocessingDirectivesCheck.cpp (hunk 2 above)
Comment (finomen) [reply clang-tools-extra/clang-tidy/daedalean/PreprocessingDirectivesCheck.cpp]: Oops, fixed. h is for C headers.

File: clang-tools-extra/clang-tidy/daedalean/PreprocessingDirectivesCheck.cpp (hunk 3 above)
Comment (finomen) [reply clang-tools-extra/clang-tidy/daedalean/PreprocessingDirectivesCheck.cpp:118]: Is there any way to use #error and not break compilation without if/ifdef/ifndef ?
For #line  T10253

Comment (aburgm) [reply clang-tools-extra/clang-tidy/daedalean/PreprocessingDirectivesCheck.cpp:118]: > Is there any way to use #error and not break compilation without if/ifdef/ifndef ?

Haha, good point... only way I can think of is through putting some rogue file in an include path

**Review by aburgm — APPROVED** [review]

**Pull request merged**